packages/jest-config/src/normalize.js

```diff
@@ -246,136 +248,76 @@ function normalize(config, argv) {
-  } else {
-    config.testRunner = resolve(config, 'testRunner', config.testRunner);
-  }
-
   if (argv.env) {
     config.testEnvironment = argv.env;
   }
+};
 
-  if (config.testEnvironment) {
-    config.testEnvironment = getTestEnvironment(config);
-  }
-
-  let babelJest;
-  if (config.transform) {
-    const toString = Object.prototype.toString;
-    if (toString.call(config.transform) !== '[object Object]') {
-      const isString = typeof config.transform === 'string';
-      const exampleTransform = isString
-        ? config.transform
-        : '<rootDir>/node_modules/babel-jest';
-      throwConfigurationError(
-        `Configuration option \`${chalk.bold('transform')}\` ` +
-        'must be an object. Use it like this:\n\n' +
-        chalk.bold(
-          '  "transform": {\n' +
-          `    "^.+\\\\.js$": "${exampleTransform}"\n` +
-          '  }'
-        )
-      );
-    }
+function normalize(config: InitialConfig, argv: Object = {}) {
+  validate(config, VALID_CONFIG, DEPRECATED_CONFIG);
 
-    const customJSPattern = Object.keys(config.transform).find(pattern => {
-      const regex = new RegExp(pattern);
-      return regex.test('a.js') || regex.test('a.jsx');
-    });
+  normalizePreprocessor(config);
+  normalizeRootDir(config);
+  normalizeMissingOptions(config);
+  normalizeArgv(config, argv);
 
-    if (customJSPattern) {
-      const jsTransformer = Resolver.findNodeModule(
-        config.transform[customJSPattern], {
-          basedir: config.rootDir,
-        },
-      );
-      if (
-        jsTransformer &&
-        jsTransformer.includes(constants.NODE_MODULES + 'babel-jest')
-      ) {
-        babelJest = jsTransformer;
-      }
-    }
-  } else {
-    babelJest = Resolver.findNodeModule('babel-jest', {
-      basedir: config.rootDir,
-    });
-    if (babelJest) {
-      config.transform = {
-        [constants.DEFAULT_JS_PATTERN]: 'babel-jest',
-      };
-    }
+  if (config.preset) {
+    config = setupPreset(config, config.preset);
   }
-
-  let polyfillPath;
-  if (babelJest) {
-    polyfillPath =
-      Resolver.findNodeModule('babel-polyfill', {
-        basedir: config.rootDir,
-      });
+  if (config.testEnvironment) {
+    config.testEnvironment = getTestEnvironment(config);
   }
 
+  const babelJest = setupBabelJest(config);
+  const newConfig = Object.assign({}, DEFAULT_CONFIG);
+
   Object.keys(config).reduce((newConfig, key) => {
     let value;
     switch (key) {
       case 'collectCoverageOnlyFrom':
-        value = Object.keys(config[key]).reduce((normObj, filePath) => {
-          filePath = path.resolve(
-            config.rootDir,
-            _replaceRootDirTags(config.rootDir, filePath),
-          );
-          normObj[filePath] = true;
-          return normObj;
-        }, Object.create(null));
+        value = normalizeCollectCoverageOnlyFrom(config, key);
         break;
-
       case 'setupFiles':
       case 'snapshotSerializers':
+        //$FlowFixMe
```

Comment:
oh, there were so many "weird" things going on here. I need to sort this out somehow and connect with `normalize(): Config`.